### PR TITLE
PMM-6295 don't show RUNNING status when pmm-agent is stopped

### PR DESCRIPTION
**Pull request by jprukner**
[PMM-6295](https://jira.percona.com/browse/PMM-6295)

We now set  status DONE to agent when it stops or is not running on pmm-managed startup. UNKNOWN right after it's added and when connection breaks between pmm-managed and pmm-agent.

- [ ] Build: https://github.com/Percona-Lab/pmm-submodules/pull/1744

- [ ] Links to other linked pull requests (optional).
---
- [ ] Tests passed.
- [ ] Feature build pass.
- [ ] (Re)requested review.
- [ ] Fix conflicts with target branch.
- [ ] Update API dependency.
- [ ] Update jira ticket description if needed.
- [ ] Attach screenshots/console output to confirm new behavior to jira ticket, if applicable.

### CODE REVIEW

File: services/agents/registry.go [hunk 1]
```diff
@@ -699,7 +704,48 @@ func (r *Registry) runStateChangeHandler(ctx context.Context, agent *pmmAgentInf
...
+func (r *Registry) CheckRunningAgents(ctx context.Context) error {
+	agents, err := models.FindAgents(r.db.Querier, models.AgentFilters{})
+	if err != nil {
+		return errors.Wrap(err, "failed to get all agents")
+
+	}
+	for _, agent := range agents {
+		// The agents without PMMAgentID set are PMM agents itself.
+		if agent.PMMAgentID == nil && !r.IsConnected(agent.AgentID) {
+			err = r.updateAgentStatusIncludingChildren(ctx, agent.AgentID, inventorypb.AgentStatus_DONE, nil)
```
[the remark this answers is not in the document]
Comment (jprukner) [reply services/agents/registry.go]: I now set status to UNKNOWN at pmm-managed startup as we discussed in Slack.

File: services/agents/registry.go [hunk 2]
```diff
@@ -618,7 +623,7 @@ func (r *Registry) stateChanged(ctx context.Context, req *agentpb.StateChangedRe
 		}
 
 		for _, agentID := range agentIDs {
-			if err := updateAgentStatus(ctx, tx.Querier, agentID, req.Status, req.ListenPort); err != nil {
+			if err := updateAgentStatus(ctx, tx.Querier, agentID, req.Status, &req.ListenPort); err != nil {
```
[the remark this answers is not in the document]
Comment (jprukner) [reply services/agents/registry.go:626]: done

File: services/agents/registry.go [hunk 3]
```diff
@@ -699,7 +704,48 @@ func (r *Registry) runStateChangeHandler(ctx context.Context, agent *pmmAgentInf
...
+}
+
+func (r *Registry) updateAgentStatusIncludingChildren(ctx context.Context, agentID string, status inventorypb.AgentStatus, listenPort *uint32) error {
+	agents, err := models.FindAgents(r.db.Querier, models.AgentFilters{
+		PMMAgentID: agentID,
+	})
+	if err != nil {
+		return errors.Wrap(err, "failed to get pmm-agent's subordinate agents")
+	}
+	agents = append(agents, &models.Agent{
```
[the remark this answers is not in the document]
Comment (jprukner) [reply services/agents/registry.go]: done

File: services/agents/registry.go [hunk 4]
```diff
@@ -603,7 +603,12 @@ func updateAgentStatus(ctx context.Context, q *reform.Querier, agentID string, s
 	}
 
 	agent.Status = status.String()
-	agent.ListenPort = pointer.ToUint16(uint16(listenPort))
+	if listenPort != nil {
```
[the remark this answers is not in the document]
Comment (jprukner) [reply services/agents/registry.go]: was the required test related to value->pointer change? I reverted that. Or should test something else?

[the remark this answers is not in the document]
Comment (BupycHuk) [reply services/agents/registry.go]: Yeah, it was related to value->pointer change

**Review by BupycHuk — CHANGES_REQUESTED** [review]
Tests are failing, also please check old clients `pmm-admin list` outputs if they works or not.

File: services/agents/registry.go [hunk 5]
```diff
@@ -582,12 +585,9 @@ func updateAgentStatus(ctx context.Context, q *reform.Querier, agentID string, s
 	l := logger.Get(ctx)
 	l.Debugf("updateAgentStatus: %s %s %d", agentID, status, listenPort)
 
-	agent := &models.Agent{AgentID: agentID}
+	agent := &models.Agent{AgentID: agentID, ListenPort: pointer.ToUint16(uint16(listenPort))}
```
[the remark this answers is not in the document]
Comment (jprukner) [reply services/agents/registry.go:588]: thanks for pointing out :) done

File: services/agents/registry.go [hunk 6]
```diff
@@ -699,7 +698,44 @@ func (r *Registry) runStateChangeHandler(ctx context.Context, agent *pmmAgentInf
...
-// RequestStateUpdate requests state update on pmm-agent with given ID.
+// SetAllAgentsStatusUnknown goes through all pmm-agents and sets status to UNKNOWN.
+func (r *Registry) SetAllAgentsStatusUnknown(ctx context.Context) error {
+	agents, err := models.FindAgents(r.db.Querier, models.AgentFilters{})
+	if err != nil {
+		return errors.Wrap(err, "failed to get all agents")
+
+	}
+	for _, agent := range agents {
+		// The agents without PMMAgentID set are PMM agents itself.
```
[the remark this answers is not in the document]
Comment (jprukner) [reply services/agents/registry.go:709]: So only external exporter in pull mode does not have status? Is there a docs on external exporters?

File: main.go [hunk 7]
```diff
@@ -834,5 +834,16 @@ func main() {
 		cleaner.Run(ctx, cleanInterval, cleanOlderThan)
 	}()
 
+	wg.Add(1)
```
[the remark this answers is not in the document]
Comment (jprukner) [reply main.go]: Done

File: services/agents/registry.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (jprukner) [reply services/agents/registry.go:712]: I added a check for external && pull mode. @BupycHuk Did you mean it like that?

[the remark this answers is not in the document]
Comment (jprukner) [reply services/agents/registry.go:713]: I skip no all external exporters. It seems like we don't set status for those. No matter metrics mode.

File: services/agents/registry.go [hunk 8]
```diff
@@ -699,7 +698,48 @@ func (r *Registry) runStateChangeHandler(ctx context.Context, agent *pmmAgentInf
...
+	}
+	return nil
+}
+
+func (r *Registry) updateAgentStatusForChildren(ctx context.Context, agentID string, status inventorypb.AgentStatus, listenPort uint32) error {
+	agents, err := models.FindAgents(r.db.Querier, models.AgentFilters{
+		PMMAgentID: agentID,
+	})
+	if err != nil {
+		return errors.Wrap(err, "failed to get pmm-agent's subordinate agents")
```
Comment (palash25) [inline services/agents/registry.go:729]: nit: lets keep the terminology consistent
`failed to get pmm-agent's subordinate agents` -> `failed to get pmm-agent's child agents`

**Review by palash25 — APPROVED** [review]

File: services/agents/registry.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (jprukner) [reply services/agents/registry.go:713]: Also HAProxy. Is it covered by externalexportertype?

File: models/agent_model.go [hunk 9]
```diff
@@ -170,6 +171,9 @@ func (s *Agent) BeforeInsert() error {
 	if len(s.CustomLabels) == 0 {
 		s.CustomLabels = nil
 	}
+	if s.Status == "" && s.PMMAgentID != nil {
```
Comment (jprukner) [inline models/agent_model.go:174]: check agent type here, too

File: services/agents/registry.go [hunk 10]
```diff
@@ -703,7 +702,48 @@ func (r *Registry) runStateChangeHandler(ctx context.Context, agent *pmmAgentInf
...
+		return errors.Wrap(err, "failed to get all agents")
+
+	}
+	for _, agent := range agents {
+		// Since we don't set status for external exporters we can skip them.
+		if agent.AgentType == models.ExternalExporterType {
+			continue
+		}
+		// The agents without PMMAgentID set are PMM agents itself.
+		if agent.PMMAgentID == nil && !r.IsConnected(agent.AgentID) {
```
Comment (BupycHuk) [inline services/agents/registry.go:718]:
```suggestion
		if agent.AgentType == models.PMMAgentType && !r.IsConnected(agent.AgentID) {
```

File: services/agents/registry.go [hunk 11]
```diff
@@ -703,7 +702,48 @@ func (r *Registry) runStateChangeHandler(ctx context.Context, agent *pmmAgentInf
...
+	agents, err := models.FindAgents(r.db.Querier, models.AgentFilters{})
+	if err != nil {
+		return errors.Wrap(err, "failed to get all agents")
+
+	}
+	for _, agent := range agents {
+		// Since we don't set status for external exporters we can skip them.
+		if agent.AgentType == models.ExternalExporterType {
+			continue
+		}
```
Comment (BupycHuk) [inline services/agents/registry.go:716]: it can be deleted.

File: services/agents/registry.go [hunk 12]
```diff
@@ -703,7 +702,48 @@ func (r *Registry) runStateChangeHandler(ctx context.Context, agent *pmmAgentInf
...
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (r *Registry) updateAgentStatusForChildren(ctx context.Context, agentID string, status inventorypb.AgentStatus, listenPort uint32) error {
+	agents, err := models.FindAgents(r.db.Querier, models.AgentFilters{
```
Comment (BupycHuk) [inline services/agents/registry.go:729]: let's find agents in within the transaction below.

File: models/agent_model.go [hunk 13]
```diff
@@ -170,6 +171,9 @@ func (s *Agent) BeforeInsert() error {
 	if len(s.CustomLabels) == 0 {
 		s.CustomLabels = nil
 	}
+	if s.Status == "" && s.AgentType != ExternalExporterType && s.PMMAgentID != nil {
```
Comment (BupycHuk) [inline models/agent_model.go:174]:
```suggestion
	if s.Status == "" && s.AgentType != PMMAgentType {
```

**Review by BupycHuk — CHANGES_REQUESTED** [review]

File: models/agent_model.go (hunk 13 above)
Comment (jprukner) [reply models/agent_model.go:174]: We need avoid both cases, External exporter and PMM Agent.

File: services/agents/registry.go (hunk 10 above)
Comment (jprukner) [reply services/agents/registry.go]: done

File: services/agents/registry.go (hunk 11 above)
Comment (jprukner) [reply services/agents/registry.go]: done

File: services/agents/registry.go (hunk 12 above)
Comment (jprukner) [reply services/agents/registry.go]: done

**Comment by jprukner** [comment]
tests are failing, let me fix

File: models/agent_model.go [hunk 14]
```diff
@@ -170,6 +171,9 @@ func (s *Agent) BeforeInsert() error {
 	if len(s.CustomLabels) == 0 {
 		s.CustomLabels = nil
 	}
+	if s.Status == "" && s.AgentType != ExternalExporterType && s.AgentType != PMMAgentType {
+		s.Status = inventorypb.AgentStatus_name[int32(inventorypb.AgentStatus_UNKNOWN)]
```
Comment (oter) [inline models/agent_model.go:175]: can we have a separate agent status enum on the model layer? e.g. like we have for NodeType 
https://github.com/percona/pmm-managed/blob/PMM-2.0/models/node_model.go#L33

Comment (jprukner) [reply models/agent_model.go]: I added a constant for UNKNOWN as we don't need other statuses.

File: models/agent_model.go [hunk 15]
```diff
@@ -41,6 +41,7 @@ type AgentType string
 const (
 	certificateKeyFilePlaceholder = "certificateKeyFilePlaceholder"
 	caFilePlaceholder             = "caFilePlaceholder"
+	AgentStatusUnknown            = "UNKNOWN"
```
[the remark this answers is not in the document]
Comment (BupycHuk) [reply models/agent_model.go:44]: it can be unexported

File: models/agent_helpers_test.go [hunk 16]
```diff
@@ -139,6 +139,7 @@ func TestAgentHelpers(t *testing.T) {
 			CreatedAt:    now,
 			UpdatedAt:    now,
 			NodeID:       pointer.ToString("N1"),
+			Status:       "UNKNOWN",
```
Comment (BupycHuk) [inline models/agent_helpers_test.go:142]: let's use constants

File: services/agents/registry.go [hunk 17]
```diff
@@ -703,7 +702,44 @@ func (r *Registry) runStateChangeHandler(ctx context.Context, agent *pmmAgentInf
 	}
 }
 
-// RequestStateUpdate requests state update on pmm-agent with given ID.
+// SetAllAgentsStatusUnknown goes through all pmm-agents and sets status to UNKNOWN.
+func (r *Registry) SetAllAgentsStatusUnknown(ctx context.Context) error {
+	agentType := models.PMMAgentType
+	agents, err := models.FindAgents(r.db.Querier, models.AgentFilters{AgentType: &agentType})
```
Comment (BupycHuk) [inline services/agents/registry.go:708]: 👍🏼

**Review by BupycHuk — APPROVED** [review]

**Pull request merged**